libs/vkd3d/command.c

```diff
@@ -32,17 +32,20 @@ static void d3d12_command_queue_add_submission(struct d3d12_command_queue *queue
 static void d3d12_fence_inc_ref(struct d3d12_fence *fence);
 static void d3d12_fence_dec_ref(struct d3d12_fence *fence);
 
-#define MAX_BATCHED_IMAGE_BARRIERS 16
+
 struct d3d12_command_list_barrier_batch
 {
-    VkImageMemoryBarrier vk_image_barriers[MAX_BATCHED_IMAGE_BARRIERS];
+    VkImageMemoryBarrier *vk_image_barriers;
```

Comment:
I'm not sure this is a good idea. Do we really expect > 16 batching to gain anything? Heap allocations complicate things. The search algorithms where we check for overlap are quadratic in nature, so this can easily explode if we're not careful to limit it.

Reply:
For the old barrier batch the limit is still 16. For the new one, Forza Horizon can issue 100+ copies in a row so I’m using a larger batch size to maximize the benefit. For other games, the CTR batch only goes as long as the number of CTR the game does, and when it actually does lots of copies I’d expect the GPU bound gain to be more important.

As for heap allocation, they’re a little bit concerning but didn’t seem much of a problem in testing.

Reply:
Maybe we can make a small_vector type optimization. Only allocate from heap if number of elements exceeds the array limit.

Reply:
However, given that is purely a tuning thing, it might be best to leave that out of this PR. A follow-up can do size tuning for Forza.

Reply:
Not sure what you're suggesting here? Do you mean that we should leave the batch size as 16 so it can be in-place allocated?

The limits in this patch is the untouched for the memory barrier batch and unlimited for the newly added transfer batch for Forza (bounded by the 64 transfer batch limit).

Reply:
I think for this PR, it should be left as-is as an array with 16 elements. We can bump that later once this lands.

Reply:
Actually, the patch is implemented in a way that requires unbounded barrier batch size, as it interleaves call to the Vulkan copy commands and addition to the post-barrier. I can further split out the barriers to a after_ctr function, but not sure that's what we want...

Reply:
Is it possible that barrier count > copy count? I'm not sure how that is possible.

Reply:
For pre-barriers, there can be additional barriers coming from resource tracking. For post-barriers it probably won't exceed the copy count, but for the CTR batch I'm using a limit of 64 for reasons mentioned above.

Reply:
By resource tracking, do you mean `d3d12_command_list_track_resource_usage`? I think those are a separate system, and shouldn't affect barrier batching.

Reply:
Yeah, sounds right. I guess we'll get the CTR batch limit lowered to 16 too for now.

Reply:
The heap allocation commit is dropped. It's irrelevant now anyway now we're limited by the same-subresource aliasing.